File: antd-py/examples/05_pointers.py
"""Example 05: Create, read, and update mutable pointers.

Pointers are mutable references that point to other network objects.
They are owned by a keypair and can be updated by the owner.
"""

import os

from antd import AntdClient
from antd.models import PointerTarget

client = AntdClient()

# Generate a random secret key (in production, use a proper key)
secret_key = os.urandom(32).hex()

# Store some data to point to
data_v1 = client.data_put_public(b"version 1")
data_v2 = client.data_put_public(b"version 2")

# Create a pointer to v1
target_v1 = PointerTarget(kind="chunk", address=data_v1.address)
ptr = client.pointer_create(secret_key, target_v1)
print(f"Pointer created at: {ptr.address}")

# Read the pointer
pointer = client.pointer_get(ptr.address)
print(f"Points to: {pointer.target.kind} @ {pointer.target.address}")
print(f"Counter: {pointer.counter}")

# Check existence
exists = client.pointer_exists(ptr.address)
print(f"Pointer exists: {exists}")

# Update pointer to point to v2
target_v2 = PointerTarget(kind="chunk", address=data_v2.address)
client.pointer_update(secret_key, target_v2)
print("Pointer updated to v2")

# Read again to verify
pointer = client.pointer_get(ptr.address)
print(f"Now points to: {pointer.target.address}")

print("Pointer CRUD OK!")

File: antd-js/examples/05-pointers.ts
/**
 * Example 05: Create, read, and update mutable pointers.
 *
 * Pointers are mutable references that point to other network objects.
 * They are owned by a keypair and can be updated by the owner.
 */

import { randomBytes } from "node:crypto";
import { createClient, type PointerTarget } from "../src/index.js";

const client = createClient();

// Generate a random secret key (in production, use a proper key)
const secretKey = randomBytes(32).toString("hex");

// Store some data to point to
const dataV1 = await client.dataPutPublic(Buffer.from("version 1"));
const dataV2 = await client.dataPutPublic(Buffer.from("version 2"));

// Create a pointer to v1
const targetV1: PointerTarget = { kind: "chunk", address: dataV1.address };
const ptr = await client.pointerCreate(secretKey, targetV1);
console.log(`Pointer created at: ${ptr.address}`);

// Read the pointer
let pointer = await client.pointerGet(ptr.address);
console.log(`Points to: ${pointer.target.kind} @ ${pointer.target.address}`);
console.log(`Counter: ${pointer.counter}`);

// Check existence
const exists = await client.pointerExists(ptr.address);
console.log(`Pointer exists: ${exists}`);

// Update pointer to point to v2
const targetV2: PointerTarget = { kind: "chunk", address: dataV2.address };
await client.pointerUpdate(secretKey, targetV2);
console.log("Pointer updated to v2");

// Read again to verify
pointer = await client.pointerGet(ptr.address);
console.log(`Now points to: ${pointer.target.address}`);

console.log("Pointer CRUD OK!");
